Feature/keystone policy overhaul

atengler opened the pull request.

Commit:
Keystone policy module, states and grain

epcim commented:
Linking: 
https://github.com/salt-formulas/salt-formula-keystone/pull/11
https://github.com/salt-formulas/salt-formula-cinder/pull/13
https://github.com/salt-formulas/salt-formula-nova/pull/17
https://github.com/salt-formulas/salt-formula-neutron/pull/8

epcim commented:
@dukov can you comment.

dukov commented on _modules/keystone_policy.py:
```diff
@@ -0,0 +1,72 @@
+
+def rule_list(path, **kwargs):
+    try:
+        file_handle = io.open(path, 'r')
```
Using 'with open(path) as file_handler:' seems more pythonistic way and will handle file_handler.close() in finally block. Moreover io.open is an alias for open in python3

dukov commented on _modules/keystone_policy.py:
```diff
@@ -0,0 +1,72 @@
+            return ret
+        del rules[name]
+        try:
+            file_handle = io.open(path, 'w')
```
Same as above

dukov commented on _modules/keystone_policy.py:
```diff
@@ -0,0 +1,72 @@
+        del rules[name]
+        try:
+            file_handle = io.open(path, 'w')
+            file_handle.write(unicode(json.dumps(rules, indent=4)))
```
Not sure that non-ASCII rules are supported by OpenStack

dukov commented on _modules/keystone_policy.py:
```diff
@@ -0,0 +1,72 @@
+            return {name: 'Rule %s already exists and is in correct state' % name}
+        rules.update({name: rule})
+        try:
+            file_handle = io.open(path, 'w')
```
ditto

dukov commented on _modules/keystone_policy.py:
```diff
@@ -0,0 +1,72 @@
+import apt
```
Is not used anywhere

dukov commented on _states/keystone_policy.py:
```diff
@@ -8,46 +8,75 @@
+      - path: /etc/keystone/policy.json
 
 '''
+import apt
```
Is not used anywhere

dukov commented on _grains/keystone_policy.py:
```diff
@@ -0,0 +1,14 @@
+import salt.loader
+
+
+def main():
```
It's not clear how this grain can be used

atengler replied:
Because of our mine_functions settings, all of grains items get published to Salt Mine, including this one. We can later use this grain to generate matching policy file on prx nodes as part of OpenStack Dashboard configuration, this feature is WIP as of now.

atengler commented:
@dukov Check the current state of this PR pls, I addressed your comments. Also after discussion with @cznewt we think you can use these states and modules in state files inside other formulas for core OpenStack services, because it is unlikely you will ever have an environment with let's say Nova, but without Keystone. So it shouldn't be necessary to copy these into the other formulas.

dukov commented:
> you can use these states and modules in state files inside other formulas

Excellent idea! 
In this case I'd rename keystone_policy resource to something more generic like policy_rule in order to not confuse enduser

dukov reviewed (approved):
LGTM

cznewt reviewed (approved):
Done as agreed, all looks fine
